IteratorOfLongs and IteratorOfShorts

victornoel opened the pull request:
This is for #838.

This:
- adds `IteratorOfLongs` and `IteratorOfShorts`
- uses them in `IterableOfLongs` and `IterableOfShorts`
- remove some unneeded annotations in other `IteratorOfXXX` (because it was so easy to do so and blatantly incorrect)
- improve documentation for all `IteratorOfXXX` (because I spent some time trying to understand something simple: that their role was to convert from primitive to objects)
- applied the conclusion we arrived to with @krzyk in https://github.com/yegor256/cactoos/pull/832#discussion_r186566153 about the simplest way to delegate to `IteratorOfXXX` in `IterableOfXXX`

I realise this is a bit more than the job asked, but all the other changes I made, I had to make them for the code I was introducing anyway.

0crat commented:
Job [#854](https://github.com/yegor256/cactoos/issues/854) is now in scope, role is `REV`

<!-- https://www.0crat.com/footprint/C63314D6Z/1300968223, version: 1.0-SNAPSHOT, hash: b8bf2614 -->

0crat commented:
This pull request [#854](https://github.com/yegor256/cactoos/issues/854) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/C63314D6Z/1300969294); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @llorllale[/z](https://www.0crat.com/u/llorllale) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer

<!-- https://www.0crat.com/footprint/C63314D6Z/1300970301, version: 1.0-SNAPSHOT, hash: b8bf2614 -->

Vatavuk commented:
@victornoel build is not passing, there are some formatting errors

victornoel commented:
@Vatavuk sorry about that, it seems this week is not my week ;)
I fixed the problems and also rebased to comply with the latest requirements of qulice which was recently updated.
Thanks (again) for the heads-up!

Vatavuk commented on src/test/java/org/cactoos/iterator/IteratorOfLongsTest.java:
```diff
@@ -0,0 +1,75 @@
+        this.iteratorWithFetchedElements().next();
+    }
+
+    private IteratorOfLongs iteratorWithFetchedElements() {
```
@victornoel this should be static method, but better approach would be to remove it completely and move its implementation back to tests using different iterator size in each test. This would remove dependency between tests and provide more test diversity.

victornoel replied:
I agree, I also did the same for the other `IteratorOfXXX` tests (actually some of them were already as you propose).

Vatavuk commented on src/test/java/org/cactoos/iterator/IteratorOfShortsTest.java:
```diff
@@ -0,0 +1,76 @@
+        this.iteratorWithFetchedElements().next();
+    }
+
+    private IteratorOfShorts iteratorWithFetchedElements() {
```
@victornoel same comment as for `IteratorOfLongsTest`

victornoel commented:
@Vatavuk it is fixed now and rebased on master

Vatavuk commented:
@llorllale good to merge

llorllale commented on src/main/java/org/cactoos/iterator/IteratorOfLongs.java:
```diff
@@ -0,0 +1,72 @@
+ *
+ * <p>There is no thread-safety guarantee.</p>
+ *
+ * @since 0.32
```
@victornoel should be 0.34 now

llorllale commented on src/test/java/org/cactoos/iterator/IteratorOfShortsTest.java:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Tests for {@link IteratorOfShorts}.
+ *
+ * @since 0.32
```
@victornoel 0.34 now

llorllale commented:
@victornoel there are a lot of irrelevant changes in this PR. Although they're small, they're introducing too much noise. Let's keep PRs tightly focused in the future please

victornoel commented:
@llorllale duly noted for the extra changes, next time I will open issues.

The `@since` tags should be correct now, thanks

llorllale commented:
@rultor merge

rultor commented:
> @rultor merge

@llorllale Done! FYI, the full log is [here](http://www.rultor.com/t/14552-389235323) (took me 9min)

0crat commented:
@elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova) please review this job completed by @vatavuk[/z](https://www.0crat.com/u/vatavuk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

<!-- https://www.0crat.com/footprint/C63314D6Z/1351744523, version: 1.0-SNAPSHOT, hash: 91fcb4fb -->

0crat commented:
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale)

<!-- https://www.0crat.com/footprint/C63314D6Z/1351746569, version: 1.0-SNAPSHOT, hash: 91fcb4fb -->

elenavolokhova commented:
@0crat quality good

0crat commented:
Order was finished, quality is "good": +20 point(s) just awarded to @vatavuk[/z](https://www.0crat.com/u/vatavuk)

<!-- https://www.0crat.com/footprint/C63314D6Z/1361619820, version: 1.0-SNAPSHOT, hash: 91fcb4fb -->

0crat commented:
Quality review completed: +8 point(s) just awarded to @elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova)

<!-- https://www.0crat.com/footprint/C63314D6Z/1361619822, version: 1.0-SNAPSHOT, hash: 91fcb4fb -->
